pkg/sdn/plugin/controller.go

```diff
@@ -176,85 +178,84 @@ func deleteLocalSubnetRoute(device, localSubnetCIDR string) {
 	}
 }
 
-func (plugin *OsdnNode) SetupSDN() (bool, error) {
-	localSubnetCIDR, err := plugin.getLocalSubnet()
-	if err != nil {
-		return false, err
+// Detect whether we are upgrading from a pre-CNI openshift and clean up
+// interfaces and iptables rules that are no longer required
+func (plugin *OsdnNode) dockerCleanup(localSubnetCIDR string) error {
+	cleanup := false
+
+	interfaces := []string{"lbr0", "vovsbr", "vlinuxbr"}
+
+	exec := kexec.New()
+	for _, intf := range interfaces {
+		itx := ipcmd.NewTransaction(exec, intf)
+		_, err := itx.GetAddresses()
+		itx.EndTransaction()
+		if err == nil {
+			cleanup = true
+			break
+		}
+	}
+	if !cleanup {
+		return nil
+	}
+
+	// Restart docker to kill old pods and make it use docker0 again.
+	// "systemctl restart" will bail out (unnecessarily) in the
+	// OpenShift-in-a-container case, so we work around that by sending
+	// the messages by hand.
+	if err := osexec.Command("dbus-send", "--system", "--print-reply", "--reply-timeout=2000", "--type=method_call", "--dest=org.freedesktop.systemd1", "/org/freedesktop/systemd1", "org.freedesktop.systemd1.Manager.Reload"); err != nil {
+		glog.Error(err)
 	}
+	if err := osexec.Command("dbus-send", "--system", "--print-reply", "--reply-timeout=2000", "--type=method_call", "--dest=org.freedesktop.systemd1", "/org/freedesktop/systemd1", "org.freedesktop.systemd1.Manager.RestartUnit", "string:'docker.service' string:'replace'"); err != nil {
```

Comment:
So this brings back the "we restart docker but then don't wait for it to finish restarting" bug that was otherwise going to go away, at least in the upgrade case. Consider an ExponentialBackoff checking "docker ps" or something to detect when it's running again?